### PR TITLE
Correct the number of microsteps for TMC drivers

### PR DESCRIPTION
**Pull request by anttix**
- Invert motion direction for SKR 1.3 and TMC2209 combination
- Disable Z-axis timeout for Lowrider configurations
- Correct the number of microsteps for TMC drivers

### CODE REVIEW

File: src/configs/accessories/invert-all-axes [hunk 1]
```diff
@@ -0,0 +1,3 @@
+opt_set INVERT_X_DIR true
```
Comment (jeffeb3) [inline src/configs/accessories/invert-all-axes:1]: Why mess with this in the firmware? One of the motors is going to be backwards anyway, you might as well reverse the plug.

File: src/configs/accessories/no-zaxis-timeout [hunk 2]
```diff
@@ -0,0 +1 @@
+opt_set DISABLE_INACTIVE_Z "false"
```
Comment (jeffeb3) [inline src/configs/accessories/no-zaxis-timeout:1]: :100:

File: src/configs/accessories/use-32-microsteps [hunk 3]
```diff
@@ -0,0 +1,20 @@
+opt_set DEFAULT_AXIS_STEPS_PER_UNIT "{ 200, 200, 800, 200 }"
+
+# NB! The values below will only have any effect if the board is fitted with programmable drivers
```
Comment (jeffeb3) [inline src/configs/accessories/use-32-microsteps:3]: Are you interested in using this for any TMC configurations? This adds a lot of changes.

Comment (jeffeb3) [reply src/configs/accessories/use-32-microsteps:3]: I guess it doesn't matter.

File: src/configs/accessories/invert-all-axes (hunk 1 above)
Comment (jeffeb3) [reply src/configs/accessories/invert-all-axes:1]: I'll wait and hear what you have to say, but I'd like to merge everything else, but I don't really like this inverting stuff.

Comment (anttix) [reply src/configs/accessories/invert-all-axes:1]: SKR boards use JST connectors which can not be easily reversed. Given you have correct cables of-course, but dupont does not give a secure connection when pressed into JST.

Granted, one could swap Y1 and Y2 and Z1 and Z2 to reverse the direction of Y and Z, however X is not easily reversed and unlike dupont, where pins are easy to pull out and rearrange, it's a lot more work to dig pins out of a JST connector.

I noticed direction pins are inverted in Klipper's [Example SKR 1.3 config](https://github.com/KevinOConnor/klipper/blob/master/config/generic-bigtreetech-skr-v1.3.cfg). I tested it and steppers do move in the correct direction with TMC2209s. I did not try dropping in DRV8825s to see if this weird behavior where TMC2209s move in a different direction to DRV8825s is some Marlin bug and klipper folks have actually got it right or is it actually some board weirdness.

Comment (anttix) [reply src/configs/accessories/invert-all-axes:1]: One more detail. This issue does not affect SKR Pro at all. If I drop an SKR Pro board into the same machine with the same non-reversible JST connectors, both TMCs and DRVs move in the same direction with the only config change to firmware config being stepper driver type.

File: src/configs/accessories/use-32-microsteps (hunk 3 above)
Comment (anttix) [reply src/configs/accessories/use-32-microsteps:3]: If you're referring to the number of changes to Configuration_adv.h then opt_set will change this for all TMC sections at once. Not sure if it's worth attempting to be more targeted here.

File: src/configs/accessories/invert-all-axes (hunk 1 above)
Comment (anttix) [reply src/configs/accessories/invert-all-axes]: Let's move this convesation to #12 to decouple this from the rest of the changes.

I've removed this commit from the updated branch.

**Comment by V1EngineeringInc** [comment]
We have talked about the Z axis before. What is the argument for needing more than 20 minutes? If we use this option do we not take away the other option?

I don't understand what is happening with the 32 micro steps part?

**Comment by anttix** [comment]
> We have talked about the Z axis before. What is the argument for needing more than 20 minutes? If we use this option do we not take away the other option?

This is only for Lowrider. The gantry *will* fall on this machine with a vengeance and may break the bit in the process. Unlike MPCNC where if the gantry does fall, it's going to come down with a lot less force. The firmware should never let that happen even if the user leaves the machine powered on with bit hanging in the middle of the table for extended periods of time. Be it out of ignorance or whatever the reason. 

> 
> I don't understand what is happening with the 32 micro steps part?

**Comment by anttix** [comment]
> I don't understand what is happening with the 32 micro steps part?

The current configuration is plain wrong for SKR 1.3 board configuration with TMC2209. Sure, we could also do it the other way around and switch to 16 microsteps for SKR + TMC2209 combination. However this is not easy to do as it will make it different from SKR + DRV8825 combination. It's not possible right now to make the steps/mm setting different for TMC2209s and DRV8825s on the same board as we don't have a way to vary board configuration based on what "accessories" are enabled.

We could also make it so that SKR 1.3 boards use 16 microsteps, period, even with DRV8825s, now this will make these different from RAMPS configuration where we use 32 mirosteps with these drivers.

**Comment by V1EngineeringInc** [comment]
> This is only for Lowrider.

If this is the right way instead of 20 minutes it should be for all builds right, it has little effect on the MPCNC so it is easier for me for it to be the same. One less difference to try and keep straight,

I just need to try and keep things easy. If every build has different option I have to keep it all straight when answering forum questions.

The issue is the 20 minutes. I remember the last time this came up with me saying you need to be within 20minutes of your machine at all times, and keeping non moving steppers energized sounds bad to me. The other side was the steppers should not be an issue and more than 20 minutes just in case. 

The right way to use your CNC as it says in the instructions is to have it park itself and lower to Zero when it is finished. So this is just in case as it is. If the machine is holding for hours and the power goes out same result. I would rather people park it or be within 20 minutes. This discourages someone running out on the last 5 minutes of a job to go to 7-11 or some other excuse. I am extremely paranoid of people leaving the machine unsupervised.

Your opinion on this?

**Comment by anttix** [comment]
My opinion is that we should not allow Z to fall automatically. 3D printers with heavy gantries generally work the same way. I think even my Prusa keeps Z energized although the gantry is way too light to fall on its own.

I think powering off X/Y but leaving Z enegized is a good compromise. It will make it hard to "resume" a job after being away for a long period of time so it discourages such behavior. However is someone is dumb enough to leave the router running as well, at least they will not suffer a friction fire due to a timeout.

**Comment by jeffeb3** [comment]
I do like having the configs simple. This system does let us specify minute differences between configs. I think it is smart to be caustious of that at first, and also try it out a bit.

I can appreciate not wanting people to leave their machines. But I don't think that is the only case this comes into play. I worry about the case where someone is talking with a friend, or setting up the workpiece, or they are reading some tutorial.

I'm not sure I agree with the danger of leaving the steppers engaged. Maybe I just don't appreciate the issue. I let my printer print nonstop for days. Leaving the cnc with the router off may not be the smartest idea, but it is hard to measure that risk against the risk of a cnc plunging unexpectedly. I think I would prefer it to be predictable.

**Comment by V1EngineeringInc** [comment]
Okay so how do I pull that without the microsteps stuff?

**Comment by anttix** [comment]
You can cherry-pick by commit hash or I can create another PR with just this commit. Given the simplicity angle, do we want to make this the default for MPCNC as well?

**Comment by anttix** [comment]
#15 created in case we want to turn this on for all CNC machines.

**Comment by anttix** [comment]
Ok so given there is #15 now, I have updated this PR to have only the microsteps change in it.

**Comment by anttix** [comment]
Continuing microsteps discussion. We currently enable 32 microsteps for Archim1, RAMPS and SKR 1.3 e.g. boards that are often used with DRV8825s. I don't know the history here of why 32 microsteps is needed for these drivers but 2x higher step rate is known to bog down 8bit CPUs so it may not be the best choice for RAMPS to begin with. If we switch to 16 microsteps for all machines, the main downside seems to be that current RAMPS users (and SKR users with experimental firmware) have to move jumpers around.

PS! there is probably a bug already due to microstep differences: aero-extruder will override xyz microstepping values to 16.

```
$ grep -r DEFAULT_AXIS_STEPS_PER_UNIT src/*
src/configs/accessories/aero-extruder:opt_set DEFAULT_AXIS_STEPS_PER_UNIT "{ 100, 100, 400, 837 }"
...
```

**Comment by jeffeb3** [comment]
I changed the name to help me keep them straight.

Ryan, this will slightly affect the Ramps builds, because it is calling this file instead of setting the steps/mm directly. It will also change some params in the Ramps configs, but only inside the TMC ifdefs, so it won't affect the build. If someone were making their own config, and wanted to enable 32nd microstepping on a TMC configuration, this would allow that. This has my approval.

**Comment by jeffeb3** [comment]
I think that steps/mm issue is why I don't have any aero configs ATM. I did have one, where it set them in a particular order, or overwrote them, but I trashed it since we already have so many configs.

**Comment by anttix** [comment]
@jeffeb3 what's your take? Would you like to keep 32steps around or is it time to standardize on 16 microsteps? Theoretical resolution of 1/16 with 16T pulleys and 2mm belts is 0.01mm.

**Comment by jeffeb3** [comment]
The way I've thought about this in the past is, 1/16 and 1/32 are both fine, except for a few corner cases (like the allthread Z, which has mostly been eliminated). The CPU limit on Z seems to have mostly gone away, so either doesn't bother me.

I would rather use 1/16, but I would also rather people stop buying Ramps, and they just keep doing it (which is fine, it's their machine). The common ramps kits come with drv8825, and it's easier to just let them stay at 1/32 microstepping by putting the drivers in that way. I don't see a compelling reason to force other configs, like the tmcs into 1/32 microstepping.

That's all the long winded way of saying, I think we should keep the ramps config at 1/32, and keep everything else at 1/16.

I am fine with this PR, to move the configuration of microsteps into another script. 

At some point, we were talking about writing better scripts for changing the settings. At that point, we might want to edit the steps/mm a little smarter. Such as making the microstepping scale the raw value, instead of hitting it with one line. That would help us in the cases like the titan aero on ramps or on rambo. Right now, if we had a ramps aero build, I would just add in a line at the bottom to change those to the right values in the top level file.

**Comment by V1EngineeringInc** [comment]
I'm fine with it. But why, just to do 16th on all? 32 bit boards don't care and these drivers can do much larger numbers. In the end I don't mind, all but any board that will need a physical jumper change.

Ramps has to stay 32 or a very large number of people will have to pull drivers to get at jumpers and that means pissed people and a very high chance of damage....and even more pissed people. Best not to change anything on Ramps or any board that would need physical jumpers moved. Since in reality there is no difference for doing it other than nice accounting.

I use the mp3dp aero firmware on a few printers, I'm not seeing an issue?

**Comment by jeffeb3** [comment]
This PR does two things:

1. It replaces option_set DEFAULT_STEPS_PER_AXIS for use-32-microsteps. Then use-32-microsteps sets the DEFAULT_STEPS_PER_AXIS. That's basically just code organization. NBD.
2. use-32-microsteps also changes the TMC microstepping. This doesn't affect any of our builds, because we don't call use-32-microsteps on any builds that also use TMC drivers. So this has very little risk of impacting us.

The potential upsides are:
a. It keeps the ramps 32 microstepping cleaner, code wise, naming that change. That might help us in the future with different configuration.
b. It _allows_ a configuration of 32 microstepping and TMC drivers. But it doesn't create one in our builds.

**Comment by jeffeb3** [comment]
> I use the mp3dp aero firmware on a few printers, I'm not seeing an issue?

I don't think there is a configuration for aero on ramps in these builds, but if there was, I think the steps/mm on the extruder would be wrong. I'm not worried about it, because we don't have that config. If we did, I would manually fix it after the 32nd microstep script ran, and after the aero script ran.

**Comment by V1EngineeringInc** [comment]
So would it be more proper to also have a "use-16-microsteps"?

That way when setting up a new board in the future we set it to either 16 or 32, as in the config file will have a similar format?

3D printer extruders need to have the steps calibrated so that is not big deal. I get the number close then write the actual steps per mm on each printer, on a re-flash I change to the actual value.

**Comment by jeffeb3** [comment]
Sure. But I think we are splitting hairs.

I might be convinced to treat steps/mm separately, where the use-32-microsteps script sets a variable, and the aero script sets a variable, and then at the end, we call set-steps-per-mm which reads those variables and then sets the line. It may still need tuning later, but being 2x off is no fun.

**Pull request merged**